mama/c_cpp/src/c/dqstrategyplugin.c

```diff
@@ -0,0 +1,927 @@
+    }
+    else
+    {
+        mama_log (MAMA_LOG_LEVEL_FINER,
+                  "Subscription for %s not forwarded as message seqnum is before seqnum expecting", userSymbol);
+    }
+
+    /*
+       NB!!!  - can't destroy a subscription after an initial!!!!!
+       After we forward this message we need to see if we should fill from
+     the pre initial cache*/
+    if (PRE_INITIAL_SCHEME_ON_INITIAL==
+            mamaTransportImpl_getPreInitialScheme (tport))
+    {
+        if (msgType==MAMA_MSG_TYPE_INITIAL || msgType == MAMA_MSG_TYPE_BOOK_INITIAL ||
+           (mamaTransportImpl_preRecapCacheEnabled (tport) &&  (msgType == MAMA_MSG_TYPE_RECAP || msgType == MAMA_MSG_TYPE_BOOK_RECAP )))
+        {
+            applyPreInitialCache(&ctx->mDqContext, subsc);
+            /*Clear the messages - no longer needed*/
+            dqContext_clearCache(&ctx->mDqContext, 0);
+        }
+    }
+
+
+    /* Note: do NOT access the "strategyImpl" members because the subscription
+     * may have been destroyed in the callback! */
+}
+
+mama_status
+dqstrategyMamaPlugin_initHook (mamaPluginInfo* pluginInfo)
+{
+    mamaPluginImpl* pluginImpl = NULL;
+    printf("init hook called\n\n\n");
+
+    pluginImpl = mamaPlugin_findPlugin (DQSTRATEGY_PLUGIN_NAME);
+    if (!pluginImpl)
+    {
+        return MAMA_STATUS_NOT_FOUND;
+    }
+
+    *pluginInfo = pluginImpl;
+
+    mama_status status = MAMA_STATUS_OK;
+
+    return status;
+}
+
+mama_status
+dqstrategyMamaPlugin_transportPostCreateHook (mamaPluginInfo pluginInfo, mamaTransport transport)
+{
+    const char*     middleware = NULL;
+    const char*     tportName  = NULL;
+    const char*     pluginName = NULL;
+    mamaBridgeImpl* bridgeImpl = NULL;
+    char            propName[256];
+    int             dqDisabled = 0;
+    
+    mamaTransportImpl_getDQDisabled(transport, &dqDisabled);
+    printf("transport post hook called\n\n\n");
+
+    mamaPluginImpl* pluginImpl = (mamaPluginImpl*)pluginInfo;
+
+    mamaTransportImpl_getBridge (transport, &bridgeImpl);
+    middleware = bridgeImpl->bridgeGetName ();
+    
+    mamaTransport_getName (transport, &tportName);
+
+    //e.g.mama.wmw.transport.tcp_sub.mama.plugin.name_=dqpluginexample
+    snprintf(propName, sizeof (propName), "mama.%s.transport.%s.%s", middleware, tportName, DQ_PLUGIN_PROPERTY);
+    pluginName = mama_getProperty(propName);
+    
+    if (!dqDisabled && (strcmp(pluginName , "dqstrategy") == 0 || !pluginName))
+    {
+       mamaTransportImpl_setDqPlugin (transport, pluginImpl);
+
+    }
+    
+    mama_status status = MAMA_STATUS_OK;
+
+    return status;
+}
+
+mama_status
+dqstrategyMamaPlugin_transportEventHook(mamaPluginInfo pluginInfo, mamaTransport transport, int setStale, mamaTransportEvent tportEvent)
+{
+    printf("transport event hook called\n\n\n");
+    int possiblyStaleForAll;
+
+    if (!(handleEvent (pluginInfo, transport)))
+    {
+        return MAMA_STATUS_OK;
+    }
+    
+    if (setStale)
```

Comment:
This should be inside a switch statement - and you should only do anything if the tportEvent is DISCONNECT.

Also, logic isn't quite right.  Should be

if (setStale)
{
//all stuff from your else statement
}

Reply:
Actually see below, add QUALITY too